sound/soc/sof/control.c

```diff
@@ -431,24 +432,40 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 		return -EINVAL;
 	}
 
-	if (copy_from_user(cdata->data, tlvd->tlv, header.length))
+	/* Allocate temporary buffer for new cdata */
+	new_cdata = kzalloc(sizeof(struct sof_ipc_ctrl_data) +
+			    header.length, GFP_KERNEL);
+	if (!new_cdata)
+		return -ENOMEM;
+
+	/* Point scontrol temporarily to the new data for IPC send */
+	scontrol->control_data = new_cdata;
```

Comment:
Should we roll this back on a  ABI failure? I see if the TX fails it is rolled back, but not for any other errors.

Reply:
Actually we should definitely roll this back, otherwise `scontrol->control_data` points to freed memory

Reply:
@cujomalainey it looks like Seppo does assign cdata back to scontrol->control_data further down. Is that what you're talking about?

Reply:
Ahh I see what you mean. yes, error handling is incomplete here

Reply:
@cujomalainey Oops, thanks for finding this!